overlord/snapstate/backend/copydata.go

```diff
@@ -239,3 +242,82 @@ var removeIfEmpty = func(dir string) error {
 
 	return os.Remove(dir)
 }
+
+// InitExposedSnapHome creates and initializes ~/Snap/<snapName> based on the
+// specified revision. Must be called after the snap has been migrated.
+func (b Backend) InitExposedSnapHome(snapName string, rev snap.Revision) (err error) {
+	opts := &dirs.SnapDirOptions{HiddenSnapDataDir: true}
+
+	users, err := allUsers(opts)
+	if err != nil {
+		return err
+	}
+
+	for _, usr := range users {
+		uid, gid, err := osutil.UidGid(usr)
+		if err != nil {
+			return err
+		}
+
+		newUserHome := snap.UserExposedHomeDir(usr.HomeDir, snapName)
+		if err := mkdirAllChown(newUserHome, 0700, uid, gid); err != nil {
+			return fmt.Errorf("cannot create %q: %v", newUserHome, err)
+		}
+
+		userData := snap.UserDataDir(usr.HomeDir, snapName, rev, opts)
+		entries, err := ioutil.ReadDir(userData)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				// there's nothing to copy into ~/Snap/<snap> (like on a fresh install)
+				continue
+			}
+
+			return err
+		}
+		for _, e := range entries {
+			src := filepath.Join(userData, e.Name())
+			dst := filepath.Join(newUserHome, e.Name())
+
+			if err := osutil.CopyFile(src, dst, osutil.CopyFlagPreserveAll|osutil.CopyFlagSync); err != nil {
```

Comment:
Interestingly, our CopyFile seems to work with directories only with CopyFlagPreserveAll flag because then we use `cp -a` internally. Without the flag CopyFile would expect a file and fail.

Reply:
Thanks for this. I've changed the test to also have to copy a directory, to make sure that it fails if that flag is removed.